File: src/main/java/de/scandio/e4/worker/confluence/rest/RestConfluence.java
package de.scandio.e4.worker.confluence.rest;

import com.google.gson.Gson;
import de.scandio.e4.worker.interfaces.RestClient;
import org.apache.commons.codec.binary.Base64;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.*;
import org.springframework.http.client.BufferingClientHttpRequestFactory;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.SimpleClientHttpRequestFactory;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RestConfluence implements RestClient {

	private static final Logger log = LoggerFactory.getLogger(RestConfluence.class);

	private static final Gson GSON = new Gson();

	private String username;
	private String password;
	private String baseUrl;

	public RestConfluence(String baseUrl, String username, String password) {
		this.baseUrl = baseUrl;
		if (!baseUrl.endsWith("/")) {
			this.baseUrl += "/";
		}
		this.username = username;
		this.password = password;
	}

	public String findPage(String spaceKey, String title) {
		String urlAfterBaseUrl = String.format("rest/api/content?title=%s&spaceKey=%s", title, spaceKey);
		return sendGetRequestReturnBody(urlAfterBaseUrl);
	}

	public List<Long> findPages(int limit) {
		List<Long> pageIds = new ArrayList<>();
		String url = String.format("/rest/api/content?type=page&start=0&limit=%s", limit);
		String body = sendGetRequestReturnBody(url);
		Map<String, Object> response = GSON.fromJson(body, Map.class);
		List<Map> pageObjects = (ArrayList) response.get("results");
		for (Map pageObj : pageObjects) {
			long pageId = Long.parseLong((String) pageObj.get("id"));
			pageIds.add(pageId);
		}
		return pageIds;
	}

	public String createPage(String pageTitle, String spaceKey, String content, String parentPageId) {
		String bodyTemplate = "{\"type\":\"page\",\"ancestors\":[{\"id\":%s}]\"title\":\"%s\",\"space\":{\"key\":\"%s\"},\"body\":{\"storage\":{\"value\":\"%s\",\"representation\":\"storage\"}}}";
		String body = String.format(bodyTemplate, parentPageId, pageTitle, spaceKey, content);
		return sendPostRequest("content/", body);
	}

	public String createSpace(String spaceKey, String spaceName) {
		String spaceDesc = "E4 created space. Enjoy!";
//		String bodyTemplate = "{\"key\":\"%s\",\"name\":\"%s\",\"description\":{\"plain\":{\"value\":\"%s\",\"representation\":\"plain\"}},\"metadata\":{}}";
		String bodyTemplate = "{\"key\":\"TST\",\"name\":\"Example space\",\"description\":{\"plain\":{\"value\":\"This is an example space\",\"representation\":\"plain\"}},\"metadata\":{}}";
		String body = String.format(bodyTemplate, spaceKey, spaceName, spaceDesc);
		return sendPostRequest("rest/api/content/", body);
	}

	private String sendPostRequest(String urlAfterBaseUrl, String body) {
		final String url = this.baseUrl + urlAfterBaseUrl;
		RestTemplate restTemplate = new RestTemplate(new BufferingClientHttpRequestFactory(new SimpleClientHttpRequestFactory()));
		List<ClientHttpRequestInterceptor> interceptors = new ArrayList<>();
		interceptors.add(new LoggingRequestInterceptor());
		restTemplate.setInterceptors(interceptors);

		log.debug("Sending POST request {{}} with body {{}}", url, body);

		HttpHeaders headers = new HttpHeaders();
		headers.add("Authorization", getBasicAuth());
		headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);

		HttpEntity<String> request = new HttpEntity<>(body, headers);
		ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.POST, request, String.class);
		String responseText = response.getBody();
		log.debug("Response text {{}}", responseText);
		return responseText;
	}

	public int sendGetRequestReturnStatus(String urlAfterBaseUrl) {
		return sendGetRequestReturnResponse(urlAfterBaseUrl).getStatusCodeValue();
	}

	public String sendGetRequestReturnBody(String urlAfterBaseUrl) {
		ResponseEntity<String> response = sendGetRequestReturnResponse(urlAfterBaseUrl);
		String responseText = response.getBody();
		log.debug("Response text {{}}", responseText);
		return responseText;
	}

	private ResponseEntity<String> sendGetRequestReturnResponse(String urlAfterBaseUrl) {
		final String url = this.baseUrl += urlAfterBaseUrl;
		final RestTemplate restTemplate = new RestTemplate();

		log.debug("Sending GET request {{}}", url);

		HttpHeaders headers = new HttpHeaders();
		headers.add("Authorization", getBasicAuth());

		HttpEntity<String> request = new HttpEntity<>(headers);
		ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, request, String.class);
		return response;
	}

	private String getBasicAuth() {
		String plainCreds = this.username + ":" + this.password;
		byte[] plainCredsBytes = plainCreds.getBytes();
		byte[] base64CredsBytes = Base64.encodeBase64(plainCredsBytes);
		String base64Creds = new String(base64CredsBytes);
		return "Basic " + base64Creds;
	}

}

File: src/main/kotlin/de/scandio/e4/clients/rest/RestConfluence.kt
package de.scandio.e4.clients.rest

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import de.scandio.e4.worker.rest.RestAtlassian
import de.scandio.e4.worker.services.StorageService
import de.scandio.e4.worker.util.WorkerUtils
import org.apache.commons.lang3.StringUtils
import org.apache.commons.text.StringEscapeUtils
import org.slf4j.LoggerFactory

import java.util.ArrayList
import java.util.HashMap

@Suppress("UNCHECKED_CAST", "SameParameterValue")
class RestConfluence(
        storageService: StorageService?,
        baseUrl: String,
        username: String,
        password: String) : RestAtlassian(storageService, baseUrl, username, password) {

    private val log = LoggerFactory.getLogger(RestConfluence::class.java)

    private val GSON = Gson()

    private val contentIds = HashMap<String, Long>()
    private val allSpaceKeys: ArrayList<String> = arrayListOf()

    private val REST_ENTITY_SEARCH_LIMIT = 100


    private fun findPage(spaceKey: String?, title: String): String {
        val urlAfterBaseUrl = String.format("rest/api/content?title=%s&spaceKey=%s", title, spaceKey)
        log.info("[REST] obtaining page data for spaceKey {{}} and title {{}}", spaceKey, title)
        return sendGetRequestReturnBody(urlAfterBaseUrl)
    }

    private fun findChildPageIdsUseCache(spaceKey: String?, parentPageTitle: String): List<Long> {
        var ids: List<Long>? = null
        val cacheKey = this.username + ":" + spaceKey + ":" + parentPageTitle
        if (this.storageService != null) {
            ids = this.storageService!!.getIdsByKey(cacheKey)
        }

        if (ids == null) {
            val parentContentId = getContentIdUseCache(spaceKey, parentPageTitle)
            val restUrl = "rest/api/content/$parentContentId/child/page"
            log.info("[REST] obtaining child page ids for user {{}}, spaceKey {{}}, parentPage {{}}", this.username, spaceKey, parentPageTitle)
            val responseText = sendGetRequestReturnBody(restUrl)
            ids = getContentIdsFromResultsResponse(responseText)
            if (this.storageService != null) {
                this.storageService!!.setIdsForKey(cacheKey, ids)
            }
        }

        return ids
    }

    private fun findContentIdsUseCache(limit: Int, type: String?, spaceKey: String?): List<Long> {
        var ids: List<Long>? = null
        val cacheKey = this.username + ":" + type + ":" + spaceKey + ":" + limit
        if (this.storageService != null) {
            ids = this.storageService!!.getIdsByKey(cacheKey)
        }

        if (ids == null) {
            var url = "rest/api/content?start=0&limit=%s"
            if (type != null && spaceKey != null) {
                url = String.format("$url&type=%s&spaceKey=%s", limit, type, spaceKey)
            } else if (spaceKey != null) {
                url = String.format("$url&spaceKey=%s", limit, spaceKey)
            } else if (type != null) {
                url = String.format("$url&type=%s", limit, type)
            } else {
                url = String.format(url, limit)
            }

            log.info("[REST] findContentIds for user {{}}, type {{}}, spaceKey {{}}", this.username, type, spaceKey)
            val body = sendGetRequestReturnBody(url)
            ids = getContentIdsFromResultsResponse(body)
            if (this.storageService != null) {
                this.storageService!!.setIdsForKey(cacheKey, ids)
            }
        }

        return ids
    }

    private fun getContentIdsFromResultsResponse(body: String): List<Long> {
        val ids = ArrayList<Long>()
        val pageObjects = getResultListFromResponse(body)
        for (pageObj in pageObjects) {
            val id = java.lang.Long.parseLong(pageObj["id"] as String)
            ids.add(id)
        }
        return ids
    }

    private fun getContentIdFromCreateResponse(body: String): Long {
        val jsonObject = GSON.fromJson(body, JsonObject::class.java)
        return jsonObject.get("id").asLong
    }

    fun fillCachesIfEmpty() {
        findContentIdsUseCache(REST_ENTITY_SEARCH_LIMIT, null, null)
        findSpaceKeysUseCache()
    }

    fun getRandomContentId(spaceKey: String? = null, parentPageTitle: String? = null): Long {
        val contentId: Long
        if (StringUtils.isBlank(parentPageTitle)) {
            contentId = WorkerUtils.getRandomItem(findContentIdsUseCache(REST_ENTITY_SEARCH_LIMIT, null, spaceKey))
        } else {
            contentId = WorkerUtils.getRandomItem(findChildPageIdsUseCache(spaceKey, parentPageTitle!!))
        }
        return contentId
    }

    override fun getUsernames(): List<String> {
        val body = sendGetRequestReturnBody("rest/api/group/confluence-users/member")
        return getListFromResponse("username", body)
    }

    private fun <T> getListFromResponse(key: String, responseText: String): List<T> {
        val ret = ArrayList<T>()
        val maps = getResultListFromResponse(responseText)
        for (obj in maps) {
            val item = obj[key] as T
            ret.add(item)
        }
        return ret
    }

    fun createPage(spaceKey: String, pageTitle: String, content: String, parentPageTitle: String? = null): Long {
        return createContentEntity("page", spaceKey, pageTitle, content, parentPageTitle)
    }

    fun createBlogpost(spaceKey: String, title: String, content: String): Long {
        return createContentEntity("blogpost", spaceKey, title, content, null)
    }

    private fun createContentEntity(
            type: String,
            spaceKey: String,
            entityTitle: String,
            unescapedContent: String,
            parentEntityTitle: String?
    ): Long {

        val content = StringEscapeUtils.escapeJson(unescapedContent)
        val bodyObj = jsonObject("""
{
  "type":"$type",
  "title":"$entityTitle",
  "space":{
    "key":"$spaceKey"
  },
  "body":{
    "storage":{
      "value":"$content",
      "representation":"storage"
    }
  }
}""")
        if (StringUtils.isNotBlank(parentEntityTitle)) {
            val parentContentId = getContentIdUseCache(spaceKey, parentEntityTitle!!)!!
            val ancestorsObj = jsonElement("""[{"id":$parentContentId}]""")
            bodyObj.add("ancestors", ancestorsObj)
        }
//        if (labels.isNotEmpty()) {
//            val metadataObj = JsonObject()
//            val labelArr = JsonArray()
//            for (label in labels) {
//                val labelObj = JsonObject()
//                labelObj.addProperty("name", label)
//                labelArr.add(labelObj)
//            }
//            metadataObj.add("labels", labelArr)
//            bodyObj.add("metadata", metadataObj)
//        }

        val responseText = sendPostRequest("rest/api/content/", GSON.toJson(bodyObj))
        return getContentIdFromCreateResponse(responseText)
    }

    fun getContentIdUseCache(spaceKey: String?, pageTitle: String): Long? {
        val cacheKey = "$spaceKey:$pageTitle"
        var pageId: Long? = contentIds[cacheKey]
        if (pageId == null) {
            val responseText = findPage(spaceKey, pageTitle)
            val pages = getResultListFromResponse(responseText)
            pageId = java.lang.Long.parseLong(pages[0]["id"] as String)
            contentIds["$spaceKey:$pageTitle"] = pageId
        }
        return pageId
    }

    fun findSpaceKeysUseCache(): List<String> {
        if (this.allSpaceKeys.isEmpty()) {
            this.allSpaceKeys.addAll(retrieveAllSpaceKeys())
        }
        return this.allSpaceKeys
    }

    fun createSpace(spaceKey: String, spaceName: String): String {
        val spaceDesc = "Space used for testing"
        val body =
"""
{
  "key":"$spaceKey",
  "name":"$spaceName",
  "description": {
    "plain": {
      "value":"$spaceDesc",
      "representation":"plain"
    }
  },
  "metadata": {}
}
"""
        return sendPostRequest("rest/api/space/", body)
    }

    fun jsonObject(jsonString: String): JsonObject {
        return GSON.fromJson(jsonString, JsonObject::class.java)
    }

    fun jsonElement(jsonString: String): JsonElement {
        return GSON.fromJson(jsonString, JsonElement::class.java)
    }

    fun addLabelsToContentEntity(contentId: Long, labels: List<String>) {
        val labelArr = JsonArray()
        for (label in labels) {
            val labelObj = JsonObject()
            labelObj.addProperty("name", label)
            labelArr.add(labelObj)
        }
        val body = GSON.toJson(labelArr)
        sendPostRequest("rest/api/content/$contentId/label", body)
    }

    fun spaceExists(spaceKey: String): Boolean {
        return try {
            val response = sendGetRequestReturnResponse("rest/api/space?spaceKey=$spaceKey")
            response.statusCodeValue == 200 && !response.body.contains("\"results\":[]")
        } catch (e: Exception) {
            log.warn("Error sending get request for spaceExists")
            false
        }
    }

    private fun retrieveAllSpaceKeys(): List<String> {
        val response = sendGetRequestReturnResponse("rest/api/space")
        return getListFromResponse("key", response.body)
    }

}
